### PR TITLE
Add test on pending promise behavior after PeerConnection is closed

### PR DESCRIPTION
**Pull request by soareschen**
For some reason I do not understand, the webrtc-pc spec states that pending promises inside operation queues should never be resolved when a `RTCPeerConnection` object is closed. (w3c/webrtc-pc#150)

I create a file to specifically this behavior, since the descriptions in spec is a bit hard to follow.

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/227890523)

# Firefox (nightly channel) #
Testing web-platform-tests at revision 370648a7ccae71c3e1777bbcb0881ef32b3b7d1e
Using browser at version BuildID 20170501100238; SourceStamp 076a7a66096f9e8d102548397254be32eb26bc3d
Starting 10 test iterations
## Unstable results ##


|                    Test                    |                        Subtest                         |     Results     | Messages |
|--------------------------------------------|--------------------------------------------------------|-----------------|----------|
| `/webrtc/pending-promise-after-close.html` | `pending createOffer() should never resolve if closed` | **PASS: 20/10** |          |

## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5753/webrtc/pending-promise-after-close.html">/webrtc/pending-promise-after-close.html</a></summary>


|                             Subtest                             |     Results     |                  Messages                  |
|-----------------------------------------------------------------|-----------------|--------------------------------------------|
|                                                                 | ERROR           |                                            |
| `pending createOffer() should never resolve if closed`          | **PASS: 20/10** |                                            |
| `pending setLocalDescription() should never resolve if closed`  | PASS            |                                            |
| `pending setRemoteDescription() should never resolve if closed` | PASS            |                                            |
| `pending addIceCandidate() should never resolve if closed`      | FAIL            | `Not enough arguments to RTCIceCandidate.` |
| `pending createAnswer() should never resolve if closed`         | PASS            |                                            |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/227890529)

# Chrome (unstable channel) #
Testing web-platform-tests at revision 370648a7ccae71c3e1777bbcb0881ef32b3b7d1e

File: webrtc/pending-promise-after-close.html [hunk 1]
```diff
@@ -0,0 +1,64 @@
+<!doctype html>
+<html>
+<head>
+  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
```
Comment (foolip) [inline webrtc/pending-promise-after-close.html]: Just `<meta charset=UTF-8>` is enough, and omitting is OK for all-ASCII files.

File: webrtc/pending-promise-after-close.html [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  <title>Test pending promises behavior when RTCPeerConnection object is closed</title>
+</head>
+<body>
+  <!-- These files are in place when executing on W3C. -->
```
Comment (foolip) [inline webrtc/pending-promise-after-close.html]: This is an odd comment in some of the existing tests. I'd suggest not copying it around, and removing it where it exists if you like.

File: webrtc/pending-promise-after-close.html [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    test_never_resolved(pc =>
+      pc.setLocalDescription({ type: 'offer' }),
+      'pending setLocalDescription() should never resolve if closed');
+
+    test_never_resolved(pc =>
+      pc.setRemoteDescription({ type: 'offer' }),
+      'pending setRemoteDescription() should never resolve if closed');
+
+    test_never_resolved(pc =>
+      pc.addIceCandidate(new RTCIceCandidate()),
```
Comment (foolip) [inline webrtc/pending-promise-after-close.html]: Can you use an empty `{}` here, since the argument is a RTCIceCandidateInit with no required arguments?

File: webrtc/pending-promise-after-close.html [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          t.step_func(err => {
+           assert_unreached(`Pending promise should never be resolved. Instead it is rejected with: ${err}`)
+          }));
+
+        // Close the PeerConnection before enqueued operation is executed.
+        pc.close();
+
+        t.step_timeout(() => {
+          t.done();
+        }, 2000);
```
Comment (foolip) [inline webrtc/pending-promise-after-close.html]: One way to avoid timeouts in a test like this is to use another chain of events that we know would be completed after the first. So, you could create a second RTCPeerConnection where you run the same steps and then don't call `close()`. Once that is done, if the first hasn't been resolved or rejected yet, then the test can end. Probably the test would run faster, although use more resources. Whatever you think would be nicer.

Comment (soareschen) [inline webrtc/pending-promise-after-close.html]: I have modified the code to test on both cases when peer connection is opened/closed. Turns out more steps is required to make the success case pass, because some operations can only be done when peer connection is in specific state.

File: webrtc/pending-promise-after-close.html (hunk 3 above)
Comment (soareschen) [inline webrtc/pending-promise-after-close.html]: Actually I am not very sure of the correct way to correctly construct a mock `RTCIceCandidate`. The spec says that either `sdpMid` or `sdpMLineIndex` must no be `null`, but I am not sure of what mock data to put in.

Comment (foolip) [inline webrtc/pending-promise-after-close.html]: Probably the spec should change there. `new RTCIceCandidate({ candidate: ' '})` doesn't throw in Chrome or Firefox, but should per spec. But making that change would likely cause exceptions to be thrown with existing content. Can you file a spec bug?

Comment (soareschen) [inline webrtc/pending-promise-after-close.html]: It seems like current browser implementation of `RTCIceCandidate` is very way off from the spec. I have filed an issue at w3c/webrtc-pc#1165 to seek clarification first.

File: webrtc/pending-promise-after-close.html [hunk 5]
```diff
@@ -0,0 +1,81 @@
+<!doctype html>
+<title>Test pending promises behavior when RTCPeerConnection object is closed</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script type="text/javascript">
+  'use strict';
+
+  function test_never_resolved(test_func, testName) {
```
Comment (rwaldron) [inline webrtc/pending-promise-after-close.html]: Nit: mixed case? `test_func, testName`

File: webrtc/pending-promise-after-close.html [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+      const pc2 = new RTCPeerConnection();
+      const closePc2 = () => {}; // noop (don't close)
+
+      test_func(pc2, closePc2)
+      .then(
+        t.step_func_done(),
+        t.step_func(err => {
+          assert_unreached(`Operating on unclosed peer connection should succeed. Instead it is rejected with: ${err}`);
+        }));
```
Comment (rwaldron) [inline webrtc/pending-promise-after-close.html]: If this test file is for "after close", then this portion and test is invalid (L26-34).

File: webrtc/pending-promise-after-close.html [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+<title>Test pending promises behavior when RTCPeerConnection object is closed</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script type="text/javascript">
+  'use strict';
+
+  function test_never_resolved(test_func, testName) {
+    async_test(t => {
+      const pc1 = new RTCPeerConnection();
+      const closePc1 = () => pc1.close();
```
Comment (rwaldron) [inline webrtc/pending-promise-after-close.html]: Once the invalid "Operating on unclosed peer connection should succeed" test is removed, then this becomes unnecessary and all that's needed is passing the RTCPeerConnection object, which can be then call `pc.close()` directly.

File: webrtc/pending-promise-after-close.html (hunk 6 above)
Comment (soareschen) [inline webrtc/pending-promise-after-close.html]: Oh, I did this after @foolip's earlier comment that the timeout for the unreached case can be eliminated by waiting for the success case promise. So if you notice both line 23 and 31 terminate the test successfully. I can add in some comments and modify the title to make it more clear.

> One way to avoid timeouts in a test like this is to use another chain of events that we know would be completed after the first. So, you could create a second RTCPeerConnection where you run the same steps and then don't call close(). Once that is done, if the first hasn't been resolved or rejected yet, then the test can end. Probably the test would run faster, although use more resources. Whatever you think would be nicer.

File: webrtc/pending-promise-after-close.html (hunk 5 above)
Comment (soareschen) [inline webrtc/pending-promise-after-close.html]: Will fix. Was switching between the case convention of testharness and normal JS and got confused.

File: webrtc/pending-promise-after-close.html (hunk 6 above)
Comment (soareschen) [inline webrtc/pending-promise-after-close.html]: I have updated the code with some comments to clarify, and removed the timeout function waiting. If you find the test helper to be too complicated, I am happy to revert back to the old approach of using timeout and create one peer connection only.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/231813122)

# Firefox (nightly) #
Testing web-platform-tests at revision 8d1fb7292edd027802a021db8b3784a4388225e6
Using browser at version BuildID 20170512100218; SourceStamp 8a7d0b15595f9916123848ca906f29c62d4914c9
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5753/webrtc/pending-promise-after-close.html">/webrtc/pending-promise-after-close.html</a></summary>


|                            Subtest                            | Results |                                                                                                                                                                                                                                    Messages                                                                                                                                                                                                                                   |
|---------------------------------------------------------------|---------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                               | OK      |                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
| `pending createOffer() should never resolve if closed`        | FAIL    | `assert_unreached: Operating on unclosed peer connection should succeed. Instead it is rejected with: InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel. Reached unreachable code`                                                                                                                                                                                                                                 |
| `pending setLocalDescription should never resolve if closed`  | FAIL    | `assert_unreached: Pending promise should never be resolved. Instead it is rejected with: InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel. Reached unreachable code;assert_unreached: Operating on unclosed peer connection should succeed. Instead it is rejected with: InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel. Reached unreachable code` |
| `pending setRemoteDescription should never resolve if closed` | FAIL    | `assert_unreached: Pending promise should never be resolved. Instead it is rejected with: InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel. Reached unreachable code`                                                                                                                                                                                                                                             |
| `pending addIceCandidate() should never resolve if closed`    | FAIL    | `Missing required 'candidate' member of RTCIceCandidateInit.`                                                                                                                                                                                                                                                                                                                                                                                                                 |
| `pending createAnswer() should never resolve if closed`       | FAIL    | `assert_unreached: Pending promise should never be resolved. Instead it is rejected with: InternalError: Cannot create an offer with no local tracks, no offerToReceiveAudio/Video, and no DataChannel. Reached unreachable code`                                                                                                                                                                                                                                             |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/231813128)

# Chrome (unstable) #
Testing web-platform-tests at revision 8d1fb7292edd027802a021db8b3784a4388225e6

**Comment by soareschen** [comment]
Closing this with the test being merged into #5847.
